Support symbolic decomposition for state prep via alias sampling

tanujkhattar opened the pull request:
Fixes https://github.com/quantumlib/Qualtran/issues/951

In general, the cirq-interop decomposition assumes access to individual bits. So anytime we would do `bb.split` / `bb.join` in the decomposition, its fine to use the cirq style decomposition as well but if we are not doing any splits or joins (example: for higher level bloqs like state prep alias sampling) its useful to migrate the cirq style decomposition to bloq style decomposition so we can support decomposing bloqs with symbolic register sizes. 

More such migrations would follow as we discover bloqs which currently have cirq-style `decompose_from_registers` implemented but can benefit from a symbolic decomposition. An alternate strategy would be to update the cirq-interop code to support symbolics wherever we can.

mpharrigan commented on qualtran/bloqs/state_preparation/state_preparation_alias_sampling.py:
```diff
@@ -209,21 +209,31 @@ def qrom_bloq(self) -> QROM:
-        yield LessThanEqual(self.mu, self.mu).on(*keep, *sigma_mu, *less_than_equal)
-        yield CSwap.make_on(ctrl=less_than_equal, x=alt, y=selection)
+            return soqs
+        selection, less_than_equal = soqs['selection'], soqs['less_than_equal']
+        sigma_mu, alt, keep = soqs['sigma_mu'], soqs['alt'], soqs['keep']
```
please do this unpacking in the method signature

tanujkhattar replied:
It'll be annoying to create the return dictionary twice when we want to return early. When there are these many parameters, its more convenient to work with the dictionary directly ?

mpharrigan replied:
oh I see

mpharrigan replied:
do we need to support the mu = 0 case? what is mu? can you update the docstring to the standard format with an Args: section and a Registers: section. I think the existing Signature section is out-of-date

tanujkhattar replied:
Yes, `self.mu` case needs to be supported. `mu` is a parameter which depends on the precision and can be 0 for when precision is low (eg: probability_epsilon=0.5)

The docstring does seem to be updated and the signature of the bloq hasn't changed since it was written. I'll polish it a bit once I fix the failing tests.

mpharrigan replied:
what do you mean? there is no args section and the register names don't match the register names

tanujkhattar replied:
updated the docstring

mpharrigan commented on qualtran/bloqs/state_preparation/state_preparation_alias_sampling_test.py:
```diff
@@ -53,6 +53,8 @@ def test_state_prep_alias_sampling_symb():
     # Symbolic T-counts
     symb_t_counts = int(expected_t_count_expr.subs({L: N, sympy.Symbol(r"\epsilon"): epsilon}))
     np.testing.assert_allclose(concrete_t_counts, symb_t_counts, rtol=1e-4)
+    # Ensure the symbolic bloq can decomposes into a composite bloq
+    _ = bloq.decompose_bloq()
```
`qlt_testing.assert_valid_bloq_decomposition`

tanujkhattar replied:
done

anurudhp commented on qualtran/bloqs/state_preparation/state_preparation_alias_sampling_test.py:
```diff
@@ -53,6 +53,8 @@ def test_state_prep_alias_sampling_symb():
     symb_t_counts = int(expected_t_count_expr.subs({L: N, sympy.Symbol(r"\epsilon"): epsilon}))
     np.testing.assert_allclose(concrete_t_counts, symb_t_counts, rtol=1e-4)
+    # Ensure the symbolic bloq can decomposes into a composite bloq
+    assert_valid_bloq_decomposition(bloq)
```
also call `bloq_autotester` for the symbolic example?

tanujkhattar commented:
Something is going wrong with the cirq interop again due to which cirq simulation tests for reflection using prepare are failing. This is probably related to the few other places we've seen this happen recently. I'll have to spend more time to figure out exactly whats going wrong.

anurudhp commented:
Possible cause: `GreedyQubitManager`, even with `maximize_reuse=True`, does not reuse qubits in a bloq style decomposition. For example see the outputs below (for a decomp. with two chained allocate/free pairs):

Script
```py
from typing import Iterator

import cirq
from attrs import frozen
from numpy.typing import NDArray
from qualtran import Bloq, BloqBuilder, GateWithRegisters, Signature, SoquetT
from qualtran.bloqs.basic_gates import CNOT
from qualtran.cirq_interop.testing import GateHelper


@frozen
class MultiAlloc(Bloq):
    @property
    def signature(self) -> "Signature":
        return Signature.build(q=1)

    def build_composite_bloq(
        self, bb: "BloqBuilder", q: "SoquetT"
    ) -> dict[str, "SoquetT"]:
        for _ in range(2):
            a = bb.allocate(1)
            a, q = bb.add(CNOT(), ctrl=a, target=q)
            bb.free(a)

        return {"q": q}


class MultiAllocGWR(GateWithRegisters):
    @property
    def signature(self) -> "Signature":
        return Signature.build(q=1)

    def decompose_from_registers(
        self, *, context: cirq.DecompositionContext, **quregs: NDArray[cirq.Qid]
    ) -> Iterator[cirq.OP_TREE]:
        (q,) = quregs["q"]
        for _ in range(2):
            (a,) = context.qubit_manager.qalloc(1)
            yield CNOT().on(a, q)
            context.qubit_manager.qfree([a])


def visualize(bloq: Bloq | GateWithRegisters):
    print(bloq)
    gate = GateHelper(
        gate=bloq,
        context=cirq.DecompositionContext(
            cirq.GreedyQubitManager(prefix="_greedy", maximize_reuse=True)
        ),
    )
    print(gate.circuit)
    print()
    print(gate.decomposed_circuit)
    print()


if __name__ == "__main__":
    visualize(MultiAlloc())
    visualize(MultiAllocGWR())
```

output
```
MultiAlloc
q: ───q───

_greedy_0: ───alloc────@───free─────────────────────────
                       │
_greedy_1: ───alloc────┼────────────────────@───free────
                       │                    │
q: ───────────Y^-0.5───@───Y^0.5───Y^-0.5───@───Y^0.5───

MultiAllocGWR
q: ───MultiAllocGWR───

_greedy_0: ───@───@───
              │   │
q: ───────────X───X───
```

tanujkhattar commented:
Yup, this is known. See https://github.com/quantumlib/Qualtran/pull/963#issue-2301621495

tanujkhattar commented:
@mpharrigan This is ready for a review

anurudhp commented on qualtran/bloqs/state_preparation/state_preparation_alias_sampling.py:
```diff
@@ -235,21 +233,31 @@ def qrom_bloq(self) -> QROM:
+        soqs['selection'] = bb.add(
+            PrepareUniformSuperposition(self.n_coeff), target=soqs['selection']
+        )
         if self.mu == 0:
```
After the fix in #1092 I think `mu` would never be `0`. So the decomp can be simplified by unpacking the registers in the method signature

anurudhp replied:
Should also probably assert `mu > 0` in post init

tanujkhattar replied:
done

anurudhp commented on qualtran/bloqs/state_preparation/state_preparation_alias_sampling_test.py:
```diff
@@ -89,6 +89,8 @@ def test_state_prep_alias_sampling_symb():
     )
     np.testing.assert_allclose(concrete_t_counts, symb_t_counts, rtol=5e-4)
+    # Ensure the symbolic bloq can decomposes into a composite bloq
+    assert_valid_bloq_decomposition(bloq)
```
bloq autotest missing

tanujkhattar replied:
done

anurudhp reviewed (approved):
lgtm % minor fixes

mpharrigan commented:
@tanujkhattar do you want to merge this

tanujkhattar commented:
I've addressed the nits and also added an assertion to in post init to verify that we only do state prep over 1D register. Opened https://github.com/quantumlib/Qualtran/issues/1280 to track improvements
